homeassistant/components/binary_sensor/mqtt.py

```diff
@@ -92,9 +105,23 @@ def async_added_to_hass(self):
         @callback
         def state_message_received(topic, payload, qos):
             """Handle a new received MQTT state message."""
+            if self._json_attributes:
+                self._attributes = {}
+                try:
+                    json_dict = json.loads(payload)
+                    if isinstance(json_dict, dict):
+                        attrs = {k: json_dict[k] for k in
+                                 self._json_attributes & json_dict.keys()}
+                        self._attributes = attrs
+                    else:
+                        _LOGGER.warning("JSON result was not a dictionary")
+                except ValueError:
+                    _LOGGER.warning("MQTT payload could not be parsed as JSON")
+                    _LOGGER.debug("Erroneous JSON: %s", payload)
```

Comment:
don't double log (just dump the erroneous JSON)

Reply:
Majority of code is from sensor/mqtt.py, didn't change much just made sure it works within this component and it's backward compatible. 

How should the user output a debug payload and still get the less verbose normal message?

Reply:
just merge the two messages

Reply:
Fixed